File: server/battle/src/main/java/info/chenliang/moba/message/Position2d.java
package info.chenliang.moba.message;
import java.io.*;
import info.chenliang.talky.*;
public class Position2d{
public float x;
public float z;
public void serialize(DataOutputStream dos) throws Exception{
FieldMark fm = new FieldMark(1);
fm.mark(x != 0);
fm.mark(z != 0);
dos.write(fm.getData());
if(fm.isMarked(0)){
dos.writeFloat(x);
}
if(fm.isMarked(1)){
dos.writeFloat(z);
}
}
public void deserialize(DataInputStream dis) throws Exception{
FieldMark fm = new FieldMark(1);
dis.read(fm.getData());
if(fm.readMark()){
x = dis.readFloat();
}
if(fm.readMark()){
z = dis.readFloat();
}
}
}

File: client/Assets/Scripts/Message/info/chenliang/moba/message/Position2d.cs
using System;
using System.IO;
using MiscUtil;
using MiscUtil.IO;
using MiscUtil.Conversion;
using Talky;
namespace info.chenliang.moba.message{
public class Position2d{
public Single x;
public Single z;
public void Serialize(MemoryEndianBinaryWriter writer){
FieldMark fm = new FieldMark(1);
fm.Mark(x != 0);
fm.Mark(z != 0);
writer.Write(fm.GetData());
if(fm.IsMarked(0)){
writer.WriteSingle(x);
}
if(fm.IsMarked(1)){
writer.WriteSingle(z);
}
}
public void Deserialize(MemoryEndianBinaryReader reader){
FieldMark fm = new FieldMark(1);
reader.Read(fm.GetData(), 0, fm.GetData().Length);
if(fm.ReadMark()){
x = reader.ReadSingle();
}
if(fm.ReadMark()){
z = reader.ReadSingle();
}
}
}
}
